lms/djangoapps/courseware/courses.py

```diff
@@ -97,26 +97,17 @@ def get_course_with_access(user, action, course_key, depth=0, check_if_enrolled=
     course = get_course_by_id(course_key, depth=depth)
 
     if not has_access(user, action, course, course_key):
-        if check_if_enrolled and not CourseEnrollment.is_enrolled(user, course_key):
-            # If user is not enrolled, raise UserNotEnrolled exception that will
-            # be caught by middleware
-            raise UserNotEnrolled(course_key)
-
         # Deliberately return a non-specific error message to avoid
         # leaking info about access control settings
         raise Http404("Course not found.")
 
-    return course
-
+    if check_if_enrolled:
+        # Verify that the user is either enrolled in the course or a staff member.
+        # If user is not enrolled, raise UserNotEnrolled exception that will be caught by middleware.
+        if not ((user.id and CourseEnrollment.is_enrolled(user, course_key)) or has_access(user, 'staff', course)):
+            raise UserNotEnrolled(course_key)
 
-def get_opt_course_with_access(user, action, course_key):
```

Comment:
Note to reviewers: `get_opt_course_with_access` was removed since there were no callers.  Must be an old function.